### PR TITLE
Feature 1746 wavelet stat

### LINKED ISSUE #1746: Make the specification of a binary threshold in Wavelet-Stat optional.
## Describe the Enhancement ##
This issue arose via the JTTI project meeting. For this project, we need the ability to run Wavelet-Stat without applying a categorical threshold to convert the data to a binary image instead. This task is to make the thresholding step optional and deal with any consequences of that change.

Be sure to:
(1) Make the Wavelet-Stat thresholding step optional in the config file
(2) Consider the scientific impact of this change. For example, this may make MSE no longer summable across scales.
(3) Consider the impacts to the PostScript and NetCDF output from Wavelet-Stat.
(4) Add new unit tests to exercise this functionality.
(5) Update the scientific and technical documentation in the MET User's Guide.

Recommend coordinating with May Wong to test the changes. 

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
164951 - NOAA JTTI FV3SAR

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required: John HG
- [x] Select **scientist(s)** or **no scientist** required: Eric G

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Review **projects** and select relevant **Repository** and **Organization** ones or add "alert:NEED PROJECT ASSIGNMENT" label
- [x] Select **milestone** to next major version milestone or "Future Versions"

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## Enhancement Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)**, **Project(s)**, **Milestone**, and **Linked issues**
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by sethlinden**
## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
During development tested with a command line derived from unit test: 

./wavelet_stat /d1/projects/MET/MET_test_data/unit_test/model_data/grib1/nam_st4/nam_2012040900_F012_gSt4.grib /d1/personal/linden/git/MET/test_output/pcp_combine/stage4_2012040912_F012_APCP12.nc ./WaveletStatConfig -outdir /d1/personal/linden/git/MET/test_output/wavelet_stat -v 1

After development, update the unit test and ran the unit test
git/MET/test/xml/unit_wavelet_stat.xml

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:
Run the unit test:
git/MET/test/xml/unit_wavelet_stat.xml

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes or No]**
Yes, the documentation was also updated:
git/MET/met/docs/Users_Guide/wavelet-stat.rst

- [x] Do these changes include sufficient testing updates? **[Yes or No]**
Yes

- [x] Will this PR result in changes to the test suite? **[Yes or No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>
Yes/No. Added new test to unit test: unit_wavelet_stat.xml

- [x] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [x] After submitting the PR, select **Linked issues** with the original issue number.
- [x] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [x] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: met/src/tools/core/wavelet_stat/wavelet_stat.cc [hunk 1]
```diff
@@ -992,10 +995,22 @@ void do_intensity_scale(const NumArray &f_na, const NumArray &o_na,
 
    // Apply each threshold
    for(i=0; i<conf_info.fcat_ta[i_vx].n_elements(); i++) {
-
+
```
Comment (JohnHalleyGotway) [inline met/src/tools/core/wavelet_stat/wavelet_stat.cc:998]:
```suggestion

```

File: met/src/tools/core/wavelet_stat/wavelet_stat.cc [hunk 2]
```diff
@@ -992,10 +995,22 @@ void do_intensity_scale(const NumArray &f_na, const NumArray &o_na,
...
    // Apply each threshold
    for(i=0; i<conf_info.fcat_ta[i_vx].n_elements(); i++) {
-
+      
       fcst_thresh_str = isc_info[i].fthresh.get_abbr_str();
       obs_thresh_str  = isc_info[i].othresh.get_abbr_str();
 
+      // If the forecast threshold is set to NA skip masking below
+      if(isc_info[i].fthresh.get_type() == thresh_na) {
+	mlog << Debug(2) << "The forecast threshold is NA (missing), skipping applying threshold" << "\n";
```
Comment (JohnHalleyGotway) [inline met/src/tools/core/wavelet_stat/wavelet_stat.cc:1004]:
```suggestion
	   mlog << Debug(2) << "The forecast threshold is NA (missing), skipping applying threshold" << "\n";
```

**Comment by JohnHalleyGotway** [comment]
When thresholds are applied, the difference fields in the PostScript output are plotted using a [-1, 1] range. When thresholds are NOT applied, the difference fields are plotted using [-n, n] where n is the maximum absolute value of the raw difference field. This is a good default value since it'll always work. However, it's a bad default since most of the differences are washed out and down closer to 0.

Difference field plotted using the old default [-1, 1]:
Screen Shot 2021-07-12 at 9 55 11 AM

Difference fields plotted using the new [-104, 104]:
Screen Shot 2021-07-12 at 9 53 39 AM

So what's the best way to handle the default plotting range for difference fields?

**Review by TaraJensen — APPROVED** [review]
Documentation is much more descriptive.  Thanks! I approve.

**Review by JohnHalleyGotway — APPROVED** [review]
I approve of these changes and note that they will change the output from the unit tests as well as the Wavelet-Stat output created by running 'make test'.

File: met/docs/Users_Guide/wavelet-stat.rst [hunk 3]
```diff
@@ -230,6 +230,21 @@ _______________________
...
 .. code-block:: none
 
+  
+  cat_thresh = [];
+  cat_thresh = [>0.0, >=5.0, NA];
+
+   
+The **cat_thresh** option defines an array of thresholds for each field defined in the fcst and obs dictionaries. The number of forecast and observation categorical thresholds must match. If set to an empty list the thresholds will not be applied (no binary masking) and all the raw grid-point values will be used for downstream statistics.
+
+If the array of threhsolds is an empy list, the application will set the threshold to NA internally and skip applying the thresholds. If the threshold is set to NA explicitly in the list, the application will also set the threshold to NA internally and skip applying the thresholds.
```
Comment (JohnHalleyGotway) [inline met/docs/Users_Guide/wavelet-stat.rst:240]:
```suggestion
If the array of thresholds is an empty list, the application will set the threshold to NA internally and skip applying the thresholds. If the threshold is set to NA explicitly in the list, the application will also set the threshold to NA internally and skip applying the thresholds.
```

**Pull request merged**